pkg/cosign/cosign.go

```diff
@@ -0,0 +1,316 @@
+}
+
+var ErrAnnotationsUnmatched = cosign.NewVerificationError("annotations unmatched")
+
+func checkAnnotations(payloads []payload.SimpleContainerImage, annotationRule v1.SignatureAnnotations) error {
+
+	sel, err := annotationRule.AsSelector()
+	if err != nil {
+		return fmt.Errorf("failed to parse annotation rule: %w", err)
+	}
+
+	if sel.Empty() {
+		return nil
+	}
+
+	kvLists := map[string][]string{}
+	for _, p := range payloads {
+		for k, v := range p.Optional {
+			if v != nil {
+				kvLists[k] = append(kvLists[k], fmt.Sprint(v))
+			}
+		}
+	}
+
+	labelMaps := generateVariations(kvLists) // alternatively we can re-write the label matching logic to work with a map[string][]string
+	logrus.Debugf("Checking against %d generated label maps: %+v", len(labelMaps), labelMaps)
+
+	for _, labelMap := range labelMaps {
+		if sel.Matches(labels.Set(labelMap)) {
+			return nil
+		}
+	}
+
+	logrus.Debugf("No label map variation matched the selector %+v", sel)
+
+	return ErrAnnotationsUnmatched
+}
+
+func generateVariations(input map[string][]string) []map[string]string {
+	// Count the number of variations
+	count := 1
+	for _, values := range input {
+		count *= len(values)
+	}
+
+	// Initialize the output slice
+	output := make([]map[string]string, count)
+
+	// Generate variations
+	for i := 0; i < count; i++ {
+		output[i] = make(map[string]string)
+		j := 1
+		for key, values := range input {
+			output[i][key] = values[(i/j)%len(values)]
+			j *= len(values)
+		}
+	}
+
+	return output
+}
+
+func verifySignatures(ctx context.Context, sigs oci.Signatures, h ggcrv1.Hash, co *cosign.CheckOpts) (checkedSignatures []oci.Signature, bundleVerified bool, err error) {
+	sl, err := sigs.Get()
+	if err != nil {
+		return nil, false, err
+	}
+
+	validationErrs := []string{}
+
+	for _, sig := range sl {
+		sig, err := static.Copy(sig)
+		if err != nil {
+			validationErrs = append(validationErrs, err.Error())
+			continue
+		}
+		verified, err := cosign.VerifyImageSignature(ctx, sig, h, co)
+		bundleVerified = bundleVerified || verified
+		if err != nil {
+			validationErrs = append(validationErrs, err.Error())
+			continue
+		}
+
+		checkedSignatures = append(checkedSignatures, sig)
+	}
+	if len(checkedSignatures) == 0 {
+		return nil, false, fmt.Errorf("%w:\n%s", cosign.ErrNoMatchingSignatures, strings.Join(validationErrs, "\n "))
+	}
+	return checkedSignatures, bundleVerified, nil
+}
+
+func extractPayload(verified []oci.Signature) ([]payload.SimpleContainerImage, error) {
+	var sigPayloads []payload.SimpleContainerImage
+	for _, sig := range verified {
+		pld, err := sig.Payload()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get payload: %w", err)
+		}
+
+		sci := payload.SimpleContainerImage{}
+		if err := json.Unmarshal(pld, &sci); err != nil {
+			return nil, fmt.Errorf("error decoding the payload: %w", err)
+		}
+
+		sigPayloads = append(sigPayloads, sci)
+	}
+	return sigPayloads, nil
+}
+
+var ALGORITHMS = map[string]crypto.Hash{
```

Comment:
Should this be lowercase so it doesn't get exported?

Reply:
```suggestion
var algorithms = map[string]crypto.Hash{
```